deps/3rdparty/libdispatch/src/queue.c

```diff
@@ -1978,3 +2002,68 @@ dispatch_after_f(dispatch_time_t when, dispatch_queue_t queue, void *ctxt, void
 	dispatch_resume(as_do(ds));
 }
 
+static void _dispatch_queue_init_specific_list(dispatch_queue_t dq) {
+	dispatch_queue_specific_list_t dqsq;
+	dqsq = calloc(1, sizeof(struct dispatch_queue_specific_list_s));
+	TAILQ_INIT(&dqsq->contextList);
+	dq->dq_specific_q = dqsq;
+}
+
+static void _dispatch_queue_insert_specific(dispatch_queue_t dq, dispatch_queue_specific_t dqs) {
+	dispatch_queue_specific_list_t dqsl = dq->dq_specific_q;
+	dispatch_queue_specific_t var;
+
+	TAILQ_FOREACH(var, &dqsl->contextList, specific) {
+		if (var->key == dqs->key) {
+			if(var->destructor) {
+				var->destructor(var->context);
+			}
+
+			// If a context exists for this key, simply update the context on that specific
+			if(dqs->context) {
+				var->context = dqs->context;
+				var->destructor = dqs->destructor;
+			} else {
+				TAILQ_REMOVE(&dqsl->contextList, var, specific);
+				free(var);
+			}
+			free(dqs);
+			return;
+		}
+	}
+
+	TAILQ_INSERT_TAIL(&dqsl->contextList, dqs, specific);
+}
+
+void dispatch_queue_set_specific(dispatch_queue_t queue, const void *key, void *_Nullable context, dispatch_function_t _Nullable destructor) {
+	if (slowpath(!key)) {
```

Comment:
do we need to be wary of locking and threading here?

Reply:
Fixed the locking issues. A dispatch_queue_t has been added to a dispatch_queue_t with delayed initialization. The specific list is stored on this embedded queue such that this queue can be run/locked on to perform operations on the specific list. Using the barrier functions, it's guaranteed that a single operation finishes before another is performed on that particular queue.